sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -267,29 +302,81 @@ case class HashAggregateExec(
           e.aggregateFunction.asInstanceOf[DeclarativeAggregate].mergeExpressions
       }
     }
-    ctx.currentVars = bufVars ++ input
-    val boundUpdateExpr = bindReferences(updateExpr, inputAttrs)
-    val subExprs = ctx.subexpressionEliminationForWholeStageCodegen(boundUpdateExpr)
-    val effectiveCodes = subExprs.codes.mkString("\n")
-    val aggVals = ctx.withSubExprEliminationExprs(subExprs.states) {
-      boundUpdateExpr.map(_.genCode(ctx))
-    }
-    // aggregate buffer should be updated atomic
-    val updates = aggVals.zipWithIndex.map { case (ev, i) =>
+
+    if (!conf.codegenSplitAggregateFunc) {
+      ctx.currentVars = bufVars ++ input
+      val boundUpdateExpr = updateExpr.map(BindReferences.bindReference(_, inputAttrs))
+      val subExprs = ctx.subexpressionEliminationForWholeStageCodegen(boundUpdateExpr)
+      val effectiveCodes = subExprs.codes.mkString("\n")
+      val aggVals = ctx.withSubExprEliminationExprs(subExprs.states) {
+        boundUpdateExpr.map(_.genCode(ctx))
+      }
+      // aggregate buffer should be updated atomic
+      val updates = aggVals.zipWithIndex.map { case (ev, i) =>
+        s"""
+           | ${bufVars(i).isNull} = ${ev.isNull};
+           | ${bufVars(i).value} = ${ev.value};
+       """.stripMargin
+      }
+      s"""
+         | // do aggregate
+         | // common sub-expressions
+         | $effectiveCodes
+         | // evaluate aggregate function
+         | ${evaluateVariables(aggVals)}
+         | // update aggregation buffer
+         | ${updates.mkString("\n").trim}
+     """.stripMargin
+    } else {
+      // We need to copy the aggregation buffer to local variables first because each aggregate
```

Comment:
Could you please help elaborate why copying the aggregate buffer values to local variables is required here?

In the current data layout design, `DeclarativeAggregate`s never share their buffer across aggregate expressions; so it never made any sense to me why in the old code we have to update all aggregate buffer slots at the end after all update expressions have been evaluated.

The flow of:
`Load aggregate buffer -> evaluate update expressions -> store back to aggregate buffer`
should be strictly confined to each aggregate expression.

The 3-step flow logically looks like this:
```
// step 1: load from aggregate buffer
val localAggBufSlot1_isNull = aggBufSlot1_isNull;
val localAggBufSlot1_value = aggBufSlot1_value;

// step 2: evaluate and materialize update/merge expressions
val isNull2 = false;
val value2 = (localAggBufSlot1_isNull ? 0 : localAggBufSlot1_value) + input1;

// step 3: write back to aggregate buffer
aggBufSlot1_isNull = isNull2;
aggBufSlot1_value = value2;
```

As you can see, evaluating the update/merge expressions won't have any side effects on the aggregate buffers -- they're evaluated to local variables first, and then stored back to aggregate buffers. So even if I simplify that to:
```
// NO step 1: load from aggregate buffer
// val localAggBufSlot1_isNull = aggBufSlot1_isNull;
// val localAggBufSlot1_value = aggBufSlot1_value;

// step 2: evaluate and materialize update/merge expressions, directly loading from aggregate buffers
val isNull2 = false;
val value2 = (aggBufSlot1_isNull ? 0 : aggBufSlot1_value) + input1;

// step 3: write back to aggregate buffer
aggBufSlot1_isNull = isNull2;
aggBufSlot1_value = value2;
```
It'll be just as safe.

Reply:
I think I'm starting to understand ... I had a different basic assumption from your splitting logic.

I started realizing that when reading your sample generated code in the PR description (BTW that sample code should be updated to match what it currently generates when you rebase this PR to the latest master)
```java
/* 108 */       // copy aggregation buffer to the local
/* 109 */       boolean agg_localBufIsNull = agg_bufIsNull;
/* 110 */       long agg_localBufValue = agg_bufValue;
/* 111 */       boolean agg_localBufIsNull1 = agg_bufIsNull1;
/* 112 */       double agg_localBufValue1 = agg_bufValue1;
/* 113 */       boolean agg_localBufIsNull2 = agg_bufIsNull2;
/* 114 */       long agg_localBufValue2 = agg_bufValue2;
/* 115 */       // common sub-expressions
/* 116 */
/* 117 */       // process aggregate functions to update aggregation buffer
/* 118 */       agg_doAggregateVal_coalesce(agg_localBufIsNull, agg_localBufValue, inputadapter_value, inputadapter_isNull);
/* 119 */       agg_doAggregateVal_add(agg_localBufValue1, inputadapter_isNull1, inputadapter_value1, agg_localBufIsNull1);
/* 120 */       agg_doAggregateVal_add1(inputadapter_isNull2, inputadapter_value2, agg_localBufIsNull2, agg_localBufValue2);
```
I thought for `SUM(a), AVG(a)` I was going to see `agg_doAggregateVal_sum` and `agg_doAggregateVal_avg`, but instead what I'm seeing here is a more fine-grained splitting for each update expression. So instead we get:
- `agg_doAggregateVal_coalesce` for `sum(a)`
- `agg_doAggregateVal_add` and `agg_doAggregateVal_add1` for `avg(a)`

My previous comment in this thread only applies when the splitting boundary is on per-aggregate-expression granularity, instead of on per-update-expression-in-aggregate-function granularity.

`kurtosis()` is pretty much the largest declarative aggregate function in Spark SQL right now. I don't think a single `kurtosis()` would go over 8000 bytes worth of bytecode, so maybe a per-aggregate-expression granularity would make more sense?

Reply:
Ur, I see. I never thought about that until now. plz give me some time. Based on the suggestion, I'll look for another solution.

btw, I'll update the description, too.

Reply:
@rednaxelafx Yea, your suggestion (it defines an individual function for each aggregate function) looks pretty good to me and that makes the generated code simpler. I updated the code and the description in this pr. So, could you check again when having time? Thanks!